### PR TITLE
Feat/new transaction model

### PR DESCRIPTION
**Pull request by alexgarzao**
This PR is a proposal about how can we deal with transactions, entries, and accounts, in the scenery with "high pressure".

Basically, the idea is just to have one table for all entries (like the current source), but entries don't have balance_after. Each entry has the account id, amount, operation (debit or credit), version, transaction id (to group a lot of entries), event id (to deal with idempotency), and a timestamp.

In this proposal, the balance that exists at the account table will not be updated when new transactions are created. This field was removed because it can be calculated by the query handler (thinking in a CQRS model).

Refactors:
* from entries use case to transactions use case
* NewEntry helper function to create a entry
* Enum to deal with operation type (debit or credit)
* Now Transaction is a object and has a NewTransaction help function too that validate if is a valid transaction (double entrying, balance == 0, amount always > 0, ...)
* New body when creating a transaction

JSON example used to create a transaction:

```
{
   "id":"28c547fa-4dd4-2593-945c-495678d7a123",
   "created_at":"2018-09-22T12:42:31Z",
   "entries":[
      {
         "id":"16b23084-686b-434a-8323-db483ce1e584",
         "operation":"debit",
         "account_id":"16b23084-686b-434a-8323-db483ce1e589",
         "version":0,
         "amount":123
      },
      {
         "id":"16b23084-686b-434a-8323-db483ce1e586",
         "operation":"credit",
         "account_id":"16b23084-686b-434a-8323-db483ce1e581",
         "version":0,
         "amount":123
      }
   ]
}
```

### CODE REVIEW

File: pkg/command-handler/domain/ledger/entities/transaction.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+	balance := 0
+	for _, entry := range entries {
+		if entry.Amount <= 0 {
+			return nil, ErrInvalidData
+		}
+
+		if entry.Operation == DebitOperation {
+			balance += entry.Amount
+		} else {
```
Comment (Pelielo) [inline pkg/command-handler/domain/ledger/entities/transaction.go:44]: Is it possible that a `InvalidOperation` type comes through this `else` statement?

File: pkg/gateways/http/transactions/create.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/sirupsen/logrus"
+	"github.com/stone-co/the-amazing-ledger/pkg/command-handler/domain/ledger/entities"
+)
+
+type CreateTransactionRequest struct {
+	ID        uuid.UUID `json:"id"`
+	CreatedAt time.Time `json:"created_at"`
```
Comment (Pelielo) [inline pkg/gateways/http/transactions/create.go:15]: Can a user send a custom `created_at` when creating a transaction? Shouldn't it be generated by the application?

**Comment by Pelielo** [comment]
Really nice work on the refactoring!

File: pkg/command-handler/domain/ledger/entities/transaction.go (hunk 1 above)
Comment (alexgarzao) [reply pkg/command-handler/domain/ledger/entities/transaction.go:44]: Yep, you are right.

I will fix. Thx :-)

File: pkg/gateways/http/transactions/create.go (hunk 2 above)
Comment (alexgarzao) [reply pkg/gateways/http/transactions/create.go:15]: When I used this approach, I'm thinking about the case where one request fails and will be retried for a (maybe) long time.

But, thinking a bit more now, I agree with you. The ledger must generate created_at. I will fix it...

Thx again Peli!

**Review by alexmourapb — APPROVED** [review]

**Pull request merged**